TelemetryDesignDoc.md

```diff
@@ -0,0 +1,327 @@
+      ]
+    }
+  ]
+}
+```
+
+A simple builder API, similar to the current Shuffleboard API, can allow dashboard configuration from robot code:
+
+```java
+var exampleTab = Telemetry.getDashboardTab("Drive");
+exampleTab.addWidget(new Widgets.Gyro("Heading", "drive/heading"))
+  // Graph widget ctor can be variadic to handle multiple data sources
+  .addWidget(new Widgets.Graph("Wheel Speeds", "drive/leftVelocity", "drive/rightVelocity")
+    .withXLimits(0, 10)
+    .withYLimits(-50, 50));
+```
+
+The above differs from the Shuffleboard API in that it has direct programmatic representation for widgets in the form of the widget classes (e.g. `Widget.Gyro`).  Shuffleboard only exposes a generic string-valued `withProperties` decorator, which does not provide type safety for user-specified widget properties.
+
+### Deprecate all existing SmartDashboard and Shuffleboard methods, replace with implementation-neutral equivalents
+
+With an implementation-neutral metadata spec available, telemetry publishing is no longer tightly coupled to the user's choice of dashboard.  Accordingly, there is no need to maintain separate `SmartDashboard` and `Shuffleboard` APIs.  In fact, it is not necessary for the API shape itself to contain any reference to the dashboard implementation or spec - this will be done only where it is convenient and reduces boilerplate in user code.
+
+The unified replacement API will look something like:
+
+```java
+// Simple imperative logging directly delegates to NT4
+Telemetry.publishDouble("key", value);
+// Can support dimensions through the API by delegating to metadata (C++ could use templates and be smarter)
+Telemetry.publishQuantity("key", value, "meters");
+
+// Declarative binding with suppliers/consumers
+Telemetry.publishDouble("key", () -> value);
+Telemetry.subscribeDouble("key", (double value) -> doThing(value));
+
+// Declarative binding of mutable classes similar to `putData(Sendable)` in existing impl
+// Optional final metadata string for aggregation
+var telemetryBinding = Telemetry.bind("key", telemetryNode);
+// Unbind later if needed (for cleanup etc)
+telemetryBinding.unbind();
+```
+
+Metadata for a given topic will be deep-merged into the existing metadata as they are introduced.  There does not seem to be any need for a system to remove metadata - such an interaction would be a code smell.
+
+### Replace Sendable with new, cleaner API
+
+`Sendable` should be replaced by a new interface, `TelemetryNode`.  The API shape for this will be fundamentally similar to `Sendable`:
+
+```java
+public interface TelemetryNode {
+  void bind(TelemetryBuilder builder);
+}
+```
+
+The "bind" method fulfills the role of `initSendable`, with a few key differences:
+
+Unlike `Sendable`, both `TelemetryNode` will support hierarchical nesting of implementing objects - that is, in addition to exposing binding handles for fields of primitive data type, it will allow *entire* `TelemetryNode`s to be added as children:
+
```

Comment:
We shouldn't have to add a new TelemetryNode often, still, it should be fairly easy for adding custom nodes and even similar nodes with vastly different properties that could drastically change the appearance of the (custom) widget needed to display the Node.
Examples might be LimeLightVision has very different properties than the similar functions of my team written RPi vision. Both should be easily displayed by the students once the Nodes have been defined. (I don't think LimeLightVision has even defined a Sendable example - it's just a bunch of getEntry.)
Another example is the color sensor is getting a lot of different implementations (often with co-processors) as teams struggle to work-around I2C problems so what will a Color Sensor TelemetryNode look like as each team gets its own?
We've also written our own code for a LIDAR.

Reply:
WPI classes themselves will have a canonical `TelemetryNode` implementation (similar to how they have a canonical `Sendable` implementation now) - if you want to override it you can wrap them and re-define it.  Vendor code will remain the responsibility of vendors; it's not clear to me that there's any easy way to enforce that the data shapes of WPILib classes match the data shapes of vendor classes.

Reply:
I think I understand what you mean by data shapes - what I called properties. The 3D accelerometer has a shape of 3 doubles (X, Y, Z). A color sensor can have a much different shape say 5 ints (C, R, G, B, P).

After reading your design doc I see my team has always misunderstood and misused in multiple ways TEST mode and SmartDashBoard LiveWindow. I ran some tests of TEST mode now and I see 2 majors problems such that TEST mode isn't very useful to us and the "shape" of data is a big one.

SmartDashboard has no idea where to put C, R, G, B, P on the screen in test mode and it doesn't even try to display them. It does make up a half way decent layout while not in test mode.

I was thinking the JSON metadata layer/file mentioned in this design doc was going to store this layout information but maybe that's one level of detail finer than what you had in mind.

A 1-D table would suffice for a layout of most data presented as an array if there are multiple values.

Reply:
There is definitely a need to be able to communicate complex data types in an atomic fashion (for both dashboards and coprocessors), but we've not yet been able to figure out an easy to use and performant method for teams (or even vendors) to do this, so largely we've hacked around it with arrays etc.  NT supports raw data transport, but time and space efficient serialization/deserialization into data structures is an open problem.  Most other frameworks (e.g. ROS, DDS, protobuf) do this with data description files and codegen of the serialization/deserialization bits.  NT4's metadata enables communicating the format of the data, so something like protobuf is potentially an option, but what is needed is build system support and documentation.

Reply:
> After reading your design doc I see my team has always misunderstood and misused in multiple ways TEST mode and SmartDashBoard LiveWindow. I ran some tests of TEST mode now and I see 2 majors problems such that TEST mode isn't very useful to us and the "shape" of data is a big one.
> 
> SmartDashboard has no idea where to put C, R, G, B, P on the screen in test mode and it doesn't even try to display them. It does make up a half way decent layout while not in test mode.

Note that the color sensor doesn't implement sendable, which is why it didn't appear in test mode.

The way that smartdashboard and shuffleboard currently handle the layout of data is through widgets. As an example, the ADXL_345 accelerometer defines the metadeta tSmartDashboardType as "3AxisAccelerometer". https://github.com/wpilibsuite/allwpilib/blob/1023c34b1c45598e6a925467bc79ab2b5b3d5e4f/wpilibj/src/main/java/edu/wpi/first/wpilibj/ADXL345_I2C.java#L227

In Smartdashboard, there is a ThreeAxisAccelerometer Widget that displays the data. https://github.com/wpilibsuite/SmartDashboard/blob/main/src/main/java/edu/wpi/first/smartdashboard/livewindow/elements/ThreeAxisAccelerometer.java

In Shuffleboard, similarly, the widget is here: https://github.com/wpilibsuite/shuffleboard/blob/main/plugins/base/src/main/java/edu/wpi/first/shuffleboard/plugin/base/widget/ThreeAxisAccelerometerWidget.java

Reply:
Note that, per the suggested implementation in this document, the "widget type" tag will live in the *dashboard JSON*, not in the metadata.